### PR TITLE
Redirect instead of render edit

### LINKED ISSUE #132: Can't display form if there are errors saving a stop
```
n ActionView::Template::Error occurred in bus_stops#update:

  undefined method `each' for nil:NilClass
  app/views/bus_stops/_form.haml:34:in `block in _app_views_bus_stops__form_haml__3321079524308047367_47402076882260'


-------------------------------
Request:
-------------------------------

  * URL        : https://stop-project.admin.umass.edu/bus_stops/965
  * HTTP Method: PATCH
  * IP address : 174.199.19.164
  * Parameters : {"utf8"=>"✓", "_method"=>"patch", "authenticity_token"=>"iPxLhwhq18A5S7QuEMDxn9NaimfQWxLKIY9hbsPl9OKY+OuBHwKURvz06YkJJJs3MHPC6T7nhNpQDsccqOOjSQ==", "bus_stop"=>{"garage_responsible"=>"SATCo", "state_road"=>"true", "need_work"=>"1 - No issues", "sign_type"=>"Axehead (2014+)", "mounting"=>"Utility pole", "shared_sign_post_frta"=>"", "mounting_direction"=>"Away from street", "mounting_clearance"=>"60-83 inches", "bolt_on_base"=>"false", "stop_sticker"=>"No sticker", "route_stickers"=>"No stickers", "shelter"=>"No shelter", "shelter_type"=>"No shelter", "shelter_ada_compliant"=>"false", "shelter_condition"=>"No shelter", "shelter_pad_condition"=>"No shelter", "shelter_pad_material"=>"No shelter", "bench"=>"None", "bike_rack"=>"None", "schedule_holder"=>"None", "system_map_exists"=>"", "trash"=>"", "accessible"=>"true", "curb_cut"=>"Within 20 feet", "sidewalk_width"=>"More than 36 inches", "bus_pull_out_exists"=>"false", "ada_landing_pad"=>"true", "obstructions"=>"No", "lighting"=>"Within 20 feet", "solar_lighting"=>"false", "has_power"=>"", "real_time_information"=>"No", "completed"=>"1"}, "commit"=>"Save stop", "controller"=>"bus_stops", "action"=>"update", "id"=>"965"}
  * Timestamp  : 2018-08-24 10:54:37 -0400
  * Server : af-transit-app3.admin.umass.edu
  * Rails root : /srv/stop-project/releases/20180823210331
  * Process: 24674

-------------------------------
Session:
-------------------------------

  * session id: [FILTERED]
  * data: {"session_id"=>"9b046ce6ad4619aa67ba73b97c518249",
   "warden.user.user.key"=>[[37], "$2a$10$5v13RFW977yLAEd3DwlPnu"],
   "_csrf_token"=>"EASgBhdoQ4bFv12nGeRqqOMpSI7uvJYQcYGmcmsGV6s="}

-------------------------------
Environment:
-------------------------------

  * CONTENT_LENGTH                                 : 1396
  * CONTENT_TYPE                                   : application/x-www-form-urlencoded
  * HTTPS                                          : on
  * HTTP_ACCEPT                                    : text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
  * HTTP_ACCEPT_ENCODING                           : br, gzip, deflate
  * HTTP_ACCEPT_LANGUAGE                           : en-us
  * HTTP_COOKIE                                    : _stop-project_session=ZG9ZeVl1OHBOZHp3bURCZ0tRZ3BlaWZMNUZ2ZW85NjNqYVZPaENtcTF5RmV2VUZjQk9LcUFWYXNxemRINjk1aWVLcHdocE5OWVVKTXBpeXRVU2NDcG1GSzBsY2xtTURCZ1pPd1hMNGdLL1RqT3ErVG4xYUszOGM1UFZYN0pxQk9Uam5wcTZGYVJvbllOMmxrajBaRTZ4MFdtRHVKY1kvVkU2U3VleGkzU25jMEh3SGM2d09UWlA3VE52QnkrMk9sWTVuMml0NWJXWVlESldOQjd2ZFVVaDk0bStlOVIweUJHSFBINkpXejhNVT0tLWFOTllsS2hnU0NCdUNEcXczYi9JMGc9PQ%3D%3D--7ef38b038098c12a4dc113ec554621ba0683ba8b
  * HTTP_DNT                                       : 1
  * HTTP_HOST                                      : stop-project.admin.umass.edu
  * HTTP_ORIGIN                                    : https://stop-project.admin.umass.edu
  * HTTP_REFERER                                   : https://stop-project.admin.umass.edu/bus_stops/965/edit
  * HTTP_USER_AGENT                                : Mozilla/5.0 (iPhone; CPU iPhone OS 11_4_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.0 Mobile/15E148 Safari/604.1
  * HTTP_VERSION                                   : HTTP/1.1
  * ORIGINAL_FULLPATH                              : /bus_stops/965
  * ORIGINAL_SCRIPT_NAME                           :
  * PASSENGER_CONNECT_PASSWORD                     : [FILTERED]
  * PATH_INFO                                      : /bus_stops/965
  * QUERY_STRING                                   :
  * REMOTE_ADDR                                    : 174.199.19.164
  * REMOTE_PORT                                    : 4311
  * REQUEST_METHOD                                 : PATCH
  * REQUEST_URI                                    : /bus_stops/965
  * ROUTES_47402072927380_SCRIPT_NAME              :
  * SCRIPT_NAME                                    :
  * SECRET_KEY_BASE                                : [FILTERED]
  * SERVER_NAME                                    : stop-project.admin.umass.edu
  * SERVER_PORT                                    : 443
  * SERVER_PROTOCOL                                : HTTP/1.1
  * SERVER_SOFTWARE                                : Apache/2.4.6 (Red Hat Enterprise Linux) OpenSSL/1.0.2k-fips Phusion_Passenger/5.3.1
  * SSL_TLS_SNI                                    : stop-project.admin.umass.edu
  * UNIQUE_ID                                      : W4AcLUxQd8GQ6sooLQH51AAAAAc
  * action_controller.instance                     : #<BusStopsController:0x0056394f6eb818>
  * action_dispatch.backtrace_cleaner              : #<Rails::BacktraceCleaner:0x0056394d1eafa0>
  * action_dispatch.cookies                        : #<ActionDispatch::Cookies::CookieJar:0x0056394f6c9b00>
  * action_dispatch.cookies_digest                 :
  * action_dispatch.cookies_serializer             : json
  * action_dispatch.encrypted_cookie_salt          : encrypted cookie
  * action_dispatch.encrypted_signed_cookie_salt   : signed encrypted cookie
  * action_dispatch.http_auth_salt                 : http authentication
  * action_dispatch.key_generator                  : #<ActiveSupport::CachingKeyGenerator:0x0056394f96cce0>
  * action_dispatch.logger                         : #<ActiveSupport::Logger:0x0056394f6dadb0>
  * action_dispatch.parameter_filter               : [:password, :secret, :password]
  * action_dispatch.redirect_filter                : []
  * action_dispatch.remote_ip                      : 174.199.19.164
  * action_dispatch.request.content_type           : application/x-www-form-urlencoded
  * action_dispatch.request.flash_hash             : #<ActionDispatch::Flash::FlashHash:0x0056394f5dcc38>
  * action_dispatch.request.formats                : [#<Mime::Type:0x0056394d52cb00 @synonyms=["application/xhtml+xml"], @symbol=:html, @string="text/html", @hash=3081580463336955014>]
  * action_dispatch.request.parameters             : {"utf8"=>"✓", "_method"=>"patch", "authenticity_token"=>"iPxLhwhq18A5S7QuEMDxn9NaimfQWxLKIY9hbsPl9OKY+OuBHwKURvz06YkJJJs3MHPC6T7nhNpQDsccqOOjSQ==", "bus_stop"=>{"garage_responsible"=>"SATCo", "state_road"=>"true", "need_work"=>"1 - No issues", "sign_type"=>"Axehead (2014+)", "mounting"=>"Utility pol...
  * action_dispatch.request.path_parameters        : {:controller=>"bus_stops", :action=>"update", :id=>"965"}
  * action_dispatch.request.query_parameters       : {}
  * action_dispatch.request.request_parameters     : {"utf8"=>"✓", "_method"=>"patch", "authenticity_token"=>"iPxLhwhq18A5S7QuEMDxn9NaimfQWxLKIY9hbsPl9OKY+OuBHwKURvz06YkJJJs3MHPC6T7nhNpQDsccqOOjSQ==", "bus_stop"=>{"garage_responsible"=>"SATCo", "state_road"=>"true", "need_work"=>"1 - No issues", "sign_type"=>"Axehead (2014+)", "mounting"=>"Utility pol...
  * action_dispatch.request.unsigned_session_cookie: {"session_id"=>"9b046ce6ad4619aa67ba73b97c518249", "warden.user.user.key"=>[[37], "$2a$10$5v13RFW977yLAEd3DwlPnu"], "_csrf_token"=>"EASgBhdoQ4bFv12nGeRqqOMpSI7uvJYQcYGmcmsGV6s="}
  * action_dispatch.request_id                     : 39492a54-45e3-4217-a0b1-d2c59458ffd6
  * action_dispatch.routes                         : #<ActionDispatch::Routing::RouteSet:0x0056394f6fd928>
  * action_dispatch.secret_key_base                : [FILTERED]
  * action_dispatch.secret_token                   : [FILTERED]
  * action_dispatch.show_detailed_exceptions       : false
  * action_dispatch.show_exceptions                : true
  * action_dispatch.signed_cookie_salt             : signed cookie
  * exception_notifier.exception_data              : {}
  * rack.errors                                    : #<IO:0x0056394ca3db40>
  * rack.hijack                                    : #<Proc:0x0056394f73a4e0@/usr/share/ruby/vendor_ruby/phusion_passenger/rack/thread_handler_extension.rb:84 (lambda)>
  * rack.hijack?                                   : true
  * rack.input                                     : #<PhusionPassenger::Utils::TeeInput:0x0056394f73ab20>
  * rack.methodoverride.original_method            : POST
  * rack.multiprocess                              : true
  * rack.multithread                               : false
  * rack.request.cookie_hash                       : {"_stop-project_session"=>"ZG9ZeVl1OHBOZHp3bURCZ0tRZ3BlaWZMNUZ2ZW85NjNqYVZPaENtcTF5RmV2VUZjQk9LcUFWYXNxemRINjk1aWVLcHdocE5OWVVKTXBpeXRVU2NDcG1GSzBsY2xtTURCZ1pPd1hMNGdLL1RqT3ErVG4xYUszOGM1UFZYN0pxQk9Uam5wcTZGYVJvbllOMmxrajBaRTZ4MFdtRHVKY1kvVkU2U3VleGkzU25jMEh3SGM2d09UWlA3VE52QnkrMk9sWTVuMml0NWJXWVlES...
  * rack.request.cookie_string                     : _stop-project_session=ZG9ZeVl1OHBOZHp3bURCZ0tRZ3BlaWZMNUZ2ZW85NjNqYVZPaENtcTF5RmV2VUZjQk9LcUFWYXNxemRINjk1aWVLcHdocE5OWVVKTXBpeXRVU2NDcG1GSzBsY2xtTURCZ1pPd1hMNGdLL1RqT3ErVG4xYUszOGM1UFZYN0pxQk9Uam5wcTZGYVJvbllOMmxrajBaRTZ4MFdtRHVKY1kvVkU2U3VleGkzU25jMEh3SGM2d09UWlA3VE52QnkrMk9sWTVuMml0NWJXWVlESldOQjd2ZFVVaDk0bStlOVIweUJHSFBINkpXejhNVT0tLWFOTllsS2hnU0NCdUNEcXczYi9JMGc9PQ%3D%3D--7ef38b038098c12a4dc113ec554621ba0683ba8b
  * rack.request.form_hash                         : {"utf8"=>"✓", "_method"=>"patch", "authenticity_token"=>"iPxLhwhq18A5S7QuEMDxn9NaimfQWxLKIY9hbsPl9OKY+OuBHwKURvz06YkJJJs3MHPC6T7nhNpQDsccqOOjSQ==", "bus_stop"=>{"garage_responsible"=>"SATCo", "state_road"=>"true", "need_work"=>"1 - No issues", "sign_type"=>"Axehead (2014+)", "mounting"=>"Utility pol...
  * rack.request.form_input                        : #<PhusionPassenger::Utils::TeeInput:0x0056394f73ab20>
  * rack.request.form_vars                         : [FILTERED]
  * rack.request.query_hash                        : {}
  * rack.request.query_string                      :
  * rack.run_once                                  : false
  * rack.session                                   : #<ActionDispatch::Request::Session:0x0056394f6fa2c8>
  * rack.session.options                           : #<ActionDispatch::Request::Session::Options:0x0056394f6fa278>
  * rack.tempfiles                                 : []
  * rack.url_scheme                                : https
  * rack.version                                   : [1, 2]
  * warden                                         : Warden::Proxy:47402072920320 @config={:default_scope=>:user, :scope_defaults=>{}, :default_strategies=>{:user=>[:database_authenticatable]}, :intercept_401=>false, :failure_app=>#<Devise::Delegator:0x0056394f812ac0>}

-------------------------------
Backtrace:
-------------------------------

  app/views/bus_stops/_form.haml:34:in `block in _app_views_bus_stops__form_haml__3321079524308047367_47402076882260'
  app/views/bus_stops/_form.haml:7:in `_app_views_bus_stops__form_haml__3321079524308047367_47402076882260'
  app/views/bus_stops/edit.haml:5:in `_app_views_bus_stops_edit_haml___4581492320879012489_47402076875280'
  app/controllers/bus_stops_controller.rb:89:in `update'
```

Introduced by #87, when there's an error saving in `BusStopsController#update`, it renders "`edit`"

https://github.com/umts/stop-project/blob/998d09c920c59122e2bfd0547dfd741cbba6097b/app/controllers/bus_stops_controller.rb#L84-L90

But, in that case, `@fields` is undefined and the template causes an exception

https://github.com/umts/stop-project/blob/998d09c920c59122e2bfd0547dfd741cbba6097b/app/views/bus_stops/_form.haml#L34

### PR DESCRIPTION
**Pull request by jforgue**
Closes #132

### CODE REVIEW

File: spec/system/bus_stops/editing_bus_stops_spec.rb [hunk 1]
```diff
@@ -49,7 +49,7 @@
     end
     it 'stays on the edit page' do
       # url doesn't change to edit page
```
Comment (dfaulken) [inline spec/system/bus_stops/editing_bus_stops_spec.rb]: Comment is a lie now

File: app/controllers/bus_stops_controller.rb [hunk 2]
```diff
@@ -84,7 +84,10 @@ def update
       redirect_to bus_stops_path
     else
       flash[:errors] = @stop.errors.full_messages
-      render 'edit'
+      @stop.completed = false
+      @stop.completed_by = nil
+      @stop.save
```
Comment (jforgue) [inline app/controllers/bus_stops_controller.rb]: I'm open to suggestions.

Comment (dfaulken) [reply app/controllers/bus_stops_controller.rb]: I would suggest:

```ruby
@stop.update completed: false
```
Then move this line to the end of the controller method, or even as an `after_action`:

https://github.com/umts/stop-project/blob/d6ebc58a8cde7b840aa7f52f044e2d62a0a6d206/app/controllers/bus_stops_controller.rb#L81

Then the model callback will handle setting the `completed_by`:

https://github.com/umts/stop-project/blob/b83e5e2f2a8cdb39b5eeed28c0094e76c08090a7/app/models/bus_stop.rb#L242-L246

**Comment by jforgue** [comment]
Is updating the stop the best option if there are errors? What if the user then decides to not fix the errors and the stop is left with incomplete data?

**Comment by dfaulken** [comment]
I still have my original question:

> Doesn't this have the effect of throwing the staged changes out of the window?

**Comment by jforgue** [comment]
Sorry, yes it does through the staged changes out.

**Comment by jforgue** [comment]
Closing to create a better solution.

**Pull request closed without merging**